plutus-core/plutus-core/test/Evaluation/Spec.hs

```diff
@@ -1,11 +1,117 @@
-module Evaluation.Spec where
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
 
-import Evaluation.Machines (test_machines)
+module Evaluation.Spec (test_evaluation) where
+
+import PlutusCore hiding (Term)
+import PlutusCore qualified as PLC
+import PlutusCore.Builtin
+import PlutusCore.Generators (forAllNoShow, genValArg)
+import PlutusCore.Generators.AST
+import PlutusCore.Pretty
 
+import Control.Exception
+import Control.Monad.Except
+import Control.Monad.Extra
+import Data.Functor (($>))
+import Data.List.Extra qualified as List
+import Evaluation.Machines (test_machines)
+import Hedgehog hiding (Opaque, Var, eval)
+import Hedgehog.Gen qualified as Gen
+import Hedgehog.Internal.Property (failWith)
 import Test.Tasty
+import Test.Tasty.Hedgehog
+import Type.Reflection
+
+type Term = PLC.Term TyName Name DefaultUni DefaultFun ()
+
+test_builtinsDon'tThrow :: TestTree
+test_builtinsDon'tThrow =
+    testGroup
+        "Builtins don't throw"
+        $ fmap (\fun -> testProperty (display fun) $ prop_builtinsDon'tThrow fun) List.enumerate
+
+prop_builtinsDon'tThrow :: DefaultFun -> Property
```

Comment:
Haddock please - we've discussed quite a bit about what this is testing and why, let's write it down!